src/LanguageServer.php

```diff
@@ -65,6 +65,10 @@ public function __construct(ProtocolReader $reader, ProtocolWriter $writer)
     {
         parent::__construct($this, '/');
         $this->protocolReader = $reader;
+        $this->protocolReader->on('close', function () {
+            $this->shutdown();
+            $this->exit();
```

Comment:
is shutdown and exit idempotent? ie is there any issue in the case of:

1. client sends shutdown, we run `$this->shutdown()`
2. client sends exit, we run `$this->exit()`
3. client closes TCP connection
4. server close event causes us to then call `$this->shutdown()` and `$this->exit()` again.